subprojects/core/src/main/java/org/gradle/execution/taskpath/ProjectFinderByTaskPath.java

```diff
@@ -19,40 +19,123 @@
+        // We save the original path, as we are going to mangle the variable
+        String originalProjectPath = projectPath;
+        MatchedProject matchedProject;
+
         if (projectPath.startsWith(Project.PATH_SEPARATOR)) {
-            current = current.getRootProject();
+            // If a path start with a separator, we handle it as an absolute path,
+            // and get the root project instead of the current project
+            matchedProject = new MatchedProject(startProject.getRootProject(), false);
+            // As we already got the root project, we should handle the path as a relative one,
+            // by cutting the path separator off
             projectPath = projectPath.substring(1);
+        } else {
+            matchedProject = new MatchedProject(startProject, false);
         }
-        for (String pattern : projectPath.split(Project.PATH_SEPARATOR)) {
-            Map<String, Project> children = current.getChildProjects();
 
-            NameMatcher matcher = new NameMatcher();
-            Project child = matcher.find(pattern, children);
-            if (child != null) {
-                current = child;
-                continue;
+        String[] components = projectPath.split(Project.PATH_SEPARATOR);
+        for (final String component : components) {
+            // Should be checked, as split above can cause troubles,
+            // if the input is only an empty string
+            if (!component.isEmpty()) {
+                // Descend, and try to resolve the child project
+                matchedProject = resolveProject(component, matchedProject);
             }
-
-            throw new ProjectLookupException(matcher.formatErrorMessage("project", current));
+        }
+        if (matchedProject.isPatternMatched()) {
+            LOGGER.info("No exact project with name ‘{}’ has been found. Checking for abbreviated names.", originalProjectPath);
+            LOGGER.info("Found exactly one project that matches the abbreviated name ‘{}’: '{}'.", originalProjectPath, matchedProject.getProject().getPath());
+        } else {
+            LOGGER.info("Found exact project with name '{}'", matchedProject.getProject().getPath());
         }
 
-        return (ProjectInternal) current;
+        return (ProjectInternal) matchedProject.getProject();
     }
 
     public static class ProjectLookupException extends InvalidUserDataException {
         public ProjectLookupException(String message) {
             super(message);
         }
     }
+
+    /**
+     * Utility class storing information about a project resolution.
+     */
+    private static class MatchedProject {
+        /**
+         * The currently resolved project
+         */
+        private final Project project;
+
+        /**
+         * Marks if in the chain of resolutions, pattern matching was necessary
+         */
+        private final boolean patternMatched;
```

Comment:
Thank you for the good javadoc, but let's put it on the accessors (getters) and not on the fields.